pkg/securitycontextconstraints/seccomp/strategy.go

```diff
@@ -0,0 +1,176 @@
+package seccomp
+
+import (
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	podutil "k8s.io/kubernetes/pkg/api/pod"
+	api "k8s.io/kubernetes/pkg/apis/core"
+)
+
+const (
+	allowAnyProfile = "*"
+)
+
+// Strategy defines the interface for all seccomp constraint strategies.
+type SeccompStrategy interface {
+	// Generate returns a profile based on constraint rules.
+	Generate(annotations map[string]string, pod *api.Pod) (string, error)
+	// Validate ensures that the specified values fall within the range of the strategy.
+	ValidatePod(pod *api.Pod) field.ErrorList
+	// Validate ensures that the specified values fall within the range of the strategy.
+	ValidateContainer(pod *api.Pod, container *api.Container) field.ErrorList
+}
+
+type strategy struct {
+	allowedProfiles []string
+	// does the strategy allow any profile (wildcard)
+	allowAnyProfile       bool
+	runtimeDefaultAllowed bool
+}
+
+var _ SeccompStrategy = &strategy{}
+
+// NewStrategy creates a new strategy that enforces seccomp profile constraints.
+func NewSeccompStrategy(allowedProfiles []string) SeccompStrategy {
+	allowAny := false
+	allowed := make([]string, 0, len(allowedProfiles))
+	runtimeDefaultAllowed := false
+
+	for _, p := range allowedProfiles {
+		if p == allowAnyProfile {
+			allowAny = true
+			continue
+		}
+		// With the graduation of seccomp to GA we automatically convert
+		// the deprecated seccomp profile `docker/default` to `runtime/default`.
+		// This means that we now have to automatically allow `runtime/default`
+		// if a user specifies `docker/default` and vice versa in an SCC.
+		if p == v1.DeprecatedSeccompProfileDockerDefault || p == v1.SeccompProfileRuntimeDefault {
+			runtimeDefaultAllowed = true
+		}
+		allowed = append(allowed, p)
+	}
+
+	return &strategy{
+		allowedProfiles:       allowed,
+		allowAnyProfile:       allowAny,
+		runtimeDefaultAllowed: runtimeDefaultAllowed,
+	}
+}
+
+// Generate returns a profile based on constraint rules.
+func (s *strategy) Generate(podAnnotations map[string]string, pod *api.Pod) (string, error) {
+	if podAnnotations[api.SeccompPodAnnotationKey] != "" {
+		// Profile already set, nothing to do.
+		return podAnnotations[api.SeccompPodAnnotationKey], nil
+	}
+	if pod.Spec.SecurityContext != nil && pod.Spec.SecurityContext.SeccompProfile != nil {
+		// Profile field already set, translate to annotation
```

Comment:
Mind adding why to this comment?  I expected one mode of enforcement. Is this for legacy compatibility?  Annotations can be changed after pods are created, so those clients are SOL if it's used for something important.

Reply:
This is how upstream currently treats the fields, I believe it is for backward compatibility.

You can change the annotation but that shouldn't have any effect on the capabilities of an already deployed pod. Of course, if anyone wanted to check the security profile of a pod/container based on the annotation, this would be rather unreliable information.